File: src/neural_network/src/scripts/camera_usb.cpp
#include <ros/ros.h>
#include <opencv2/opencv.hpp>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/Image.h>
#include <image_transport/image_transport.h>

void find_devices()
{
    for (int i = 0; i < 5; i++)
    {
        cv::VideoCapture cap_test(i);
        if (cap_test.isOpened())
        {
            std::cout << " Device " << i << " is avaliable." << std::endl;
        }
        cap_test.release();
    }
}

int main(int argc, char **argv)
{
    ros::init(argc, argv, "camera_node");

    ros::NodeHandle nh("~"); // Crea un nodo privado

    image_transport::ImageTransport it(nh);
    image_transport::Publisher pub = it.advertise("/camera/image_raw", 1);
    find_devices();

    int deviceID, width, height;
    nh.getParam("/camera/deviceID", deviceID);
    nh.getParam("/camera/width", width);
    nh.getParam("/camera/height", height);

    cv::VideoCapture cap(0);
    if (!cap.isOpened())
        {
            ROS_ERROR("Failed to open web cam");
            return -1;
        }
    int frameWidth = width;
    cap.set(cv::CAP_PROP_FRAME_WIDTH,frameWidth);
    std::cout<< "set frame width " << frameWidth<< std::endl;

    int frameHeight = height;
    cap.set(cv::CAP_PROP_FRAME_HEIGHT,frameHeight);
    std::cout<< "set frame height " << frameHeight<< std::endl;
    std::cout << "Device ID"<< deviceID << std::endl;   

    std::cout << "Current camera properties: "<< std::endl;

    frameWidth = cap.get(cv::CAP_PROP_FRAME_WIDTH);
    std::cout<< "get frame width " << frameWidth<< std::endl;

    frameHeight = cap.get(cv::CAP_PROP_FRAME_HEIGHT);
    std::cout<< "get frame height " << frameHeight<< std::endl; 

    double frameRate = cap.get(cv::CAP_PROP_FPS);
    std::cout<< "frame " << frameRate << std::endl;

    while (ros::ok())
    {
        cv::Mat frame;
        cap >> frame;

        if (!frame.empty())
        {
            sensor_msgs::ImageConstPtr msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", frame).toImageMsg();

            pub.publish(msg);
        }
        ros::spinOnce();
    }
    return 0;
}

File: src/neural_network/src/scripts/camera_usb.py
#!/usr/bin/env python
import rospy
from cv_bridge import CvBridge, CvBridgeError
from sensor_msgs.msg import Image
import cv2

def find_devices():
    for i in range(5):
        cap_test = cv2.VideoCapture(i)
        if cap_test.isOpened():
            print("Device {} is available.".format(i))
        cap_test.release()

def main():
    rospy.init_node('camera_node', anonymous=True)
    rate = rospy.Rate(10)  # 10 Hz

    find_devices()

    device_id = rospy.get_param("/camera/deviceID", 0)
    width = rospy.get_param("/camera/width", 640)
    height = rospy.get_param("/camera/height", 480)

    cap = cv2.VideoCapture(device_id)
    if not cap.isOpened():
        rospy.logerr("Failed to open web cam")
        return

    cap.set(cv2.CAP_PROP_FRAME_WIDTH, width)
    print("Set frame width:", width)

    cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
    print("Set frame height:", height)

    print("Current camera properties:")
    print("Frame width:", cap.get(cv2.CAP_PROP_FRAME_WIDTH))
    print("Frame height:", cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
    print("Frame rate:", cap.get(cv2.CAP_PROP_FPS))

    bridge = CvBridge()
    pub = rospy.Publisher('/camera/image_raw', Image, queue_size=1)

    while not rospy.is_shutdown():
        ret, frame = cap.read()
        if ret:
            try:
                image_msg = bridge.cv2_to_imgmsg(frame, "bgr8")
                pub.publish(image_msg)
            except CvBridgeError as e:
                rospy.logerr(e)
        rate.sleep()

    cap.release()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass
